### PR TITLE
Basic conversion of `ERROR` into `dg-error` directive

### PR DESCRIPTION
**Pull request by MahadMuhammad**
A very basic PR for converting `//~^ ERROR` into `dejagnu` directive.

- Currently, I am using [`std-contains`](https://doc.rust-lang.org/std/primitive.str.html#method.contains) method for finding the `ERROR` pattern. I need to convert this into `regex` checker.
- For future, we can add [this](https://doc.rust-lang.org/cargo/guide/continuous-integration.html#github-actions) workflow for testing this tool.

##### References:
```bash
cargo run -- --file <sample-test-file>
```

##### References:
- [`ripgrep`](https://github.com/BurntSushi/ripgrep): A faster replacement for GNU’s `grep` command in rust.
- [`sd`](https://github.com/chmln/sd): Intuitive find & replace CLI
- [`rust-cli-book`](https://rust-cli.github.io/book/index.html)

---

- Sample testing on this file https://github.com/rust-lang/rust/blob/master/tests/ui/anon-params/anon-params-denied-2018.rs
<details><summary>Output: </summary>
<p>

```rust
// Tests that anonymous parameters are a hard error in edition 2018.

//@ edition:2018

trait T {
    fn foo(i32); //~ expected one of `:`, `@`, or `|`, found `)`

    // Also checks with `&`
    fn foo_with_ref(&mut i32);
    // { dg-error "expected one of `:`, `@`, or `|`, found `)`" }

    fn foo_with_qualified_path(<Bar as T>::Baz);
    // { dg-error "expected one of `(`, `...`, `..=`, `..`, `::`, `:`, `{`, or `|`, found `)`" }

    fn foo_with_qualified_path_and_ref(&<Bar as T>::Baz);
    // { dg-error "expected one of `(`, `...`, `..=`, `..`, `::`, `:`, `{`, or `|`, found `)`" }

    fn foo_with_multiple_qualified_paths(<Bar as T>::Baz, <Bar as T>::Baz);
    // { dg-error "expected one of `(`, `...`, `..=`, `..`, `::`, `:`, `{`, or `|`, found `,`" }
    //~| ERROR expected one of `(`, `...`, `..=`, `..`, `::`, `:`, `{`, or `|`, found `)`

    fn bar_with_default_impl(String, String) {}
    // { dg-error "expected one of `:`" }
    //~| ERROR expected one of `:`

    // do not complain about missing `b`
    fn baz(a:usize, b, c: usize) -> usize { //~ ERROR expected one of `:`
        a + b + c
    }
}

fn main() {}
```

</p>
</details>

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+const DG_ERROR: &str = "//~^ ERROR ";
+
+fn main() {
+    if let Err(error) = try_parse() {
+        eprintln!("Error: {:?}", error);
+        process::exit(1);
+    }
+}
```
Comment (CohenArthur) [inline src/main.rs]:
```suggestion
fn main() -> Result {
    try_parse()
}
```

File: src/main.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+fn try_parse() -> Result<()> {
+    let args = Cli::parse();
+
+    if !args.file.exists() {
+        panic!("File {:?} not found", args.file);
```
Comment (CohenArthur) [inline src/main.rs]: in a later PR: return a proper error type about how the file wasn't found/doesn't exist

File: src/main.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    let code = fs::read_to_string(&args.file).expect("Failed to read file");
```
Comment (CohenArthur) [inline src/main.rs]:
```suggestion
    let code = fs::read_to_string(&args.file)?;
```

File: src/main.rs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    let new_code = transform_code(&code, DG_ERROR);
+
+    fs::remove_file(&args.file).expect("Failed to delete file");
+    fs::write(&args.file, new_code.join("\n")).expect("Failed to write to file");
```
Comment (CohenArthur) [inline src/main.rs]:
```suggestion
    fs::remove_file(&args.file)?;
    fs::write(&args.file, new_code.join("\n"))?;
```

File: src/main.rs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+use clap::Parser;
+use std::{fs, path, process};
+
+#[derive(Parser)]
+struct Cli {
+    #[arg(short = 'f', long = "file")]
+    file: path::PathBuf,
+}
+
+const DG_ERROR: &str = "//~^ ERROR ";
```
Comment (CohenArthur) [inline src/main.rs]:
```suggestion
const RUSTTEST_ERROR: &str = "//~^ ERROR ";
```

File: src/main.rs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+fn transform_code(code: &String, dg_directive: &str) -> Vec<String> {
+    let mut new_code = Vec::new();
+
+    for line in code.lines() {
+        if line.contains(dg_directive) {
+            // replace the rust directive to dejagnu directive
+            // TODO: Add more directive relative to rustc
+            let new_line = line.replace(dg_directive, "// { dg-error \"");
```
Comment (CohenArthur) [inline src/main.rs]:
```suggestion
            let new_line = line.replace(dg_directive, DG_ERROR);
```

File: src/main.rs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+#[test]
+fn test_transform_code() {
+    let dg_msg = "// { dg-error \"expected one of `:`, `@`, or `|`, found `)`\" }";
+    let rust_msg = "//~^ ERROR expected one of `:`, `@`, or `|`, found `)`";
+    assert_eq!(
+        transform_code(&rust_msg.to_string(), DG_ERROR),
+        vec![dg_msg.to_string()]
+    );
+}
```
Comment (CohenArthur) [inline src/main.rs]:
```suggestion
#[cfg(test)]
mod tests {

use super::*;

#[test]
fn test_transform_code() {
    let dg_msg = "// { dg-error \"expected one of `:`, `@`, or `|`, found `)`\" }";
    let rust_msg = "//~^ ERROR expected one of `:`, `@`, or `|`, found `)`";
    assert_eq!(
        transform_code(&rust_msg.to_string(), DG_ERROR),
        vec![dg_msg.to_string()]
    );
}

}
```

File: src/main.rs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+    let new_code = transform_code(&code, DG_ERROR);
+
+    fs::remove_file(&args.file).expect("Failed to delete file");
+    fs::write(&args.file, new_code.join("\n")).expect("Failed to write to file");
+
+    Ok(())
+}
+
+fn transform_code(code: &String, dg_directive: &str) -> Vec<String> {
```
Comment (CohenArthur) [inline src/main.rs]:
```suggestion
fn transform_code(code: &str, dg_directive: &str) -> Vec<String> {
```

Comment (CohenArthur) [reply src/main.rs]: give `&str` or `String` - not `&String` and not `str` in most cases

File: src/main.rs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    }
+    new_code
+}
+
+#[test]
+fn test_transform_code() {
+    let dg_msg = "// { dg-error \"expected one of `:`, `@`, or `|`, found `)`\" }";
+    let rust_msg = "//~^ ERROR expected one of `:`, `@`, or `|`, found `)`";
+    assert_eq!(
+        transform_code(&rust_msg.to_string(), DG_ERROR),
```
Comment (CohenArthur) [inline src/main.rs]:
```suggestion
        transform_code(rust_msg, DG_ERROR),
```

File: src/main.rs [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+const RUSTTEST_ERROR: &str = "//~^ ERROR ";
+const DG_ERROR: &str = "// { dg-error \"";
+
+fn main() -> Result<()> {
+    if let Err(error) = try_parse() {
+        eprintln!("Error: {:?}", error);
+        process::exit(1);
+    }
+    Ok(())
```
Comment (CohenArthur) [inline src/main.rs]: so is it not possible to simplify the code here just by calling `try_parse()` and letting it return? does it produce an error?

**Review by CohenArthur — APPROVED** [review]
slight nitpick but that looks good to me :D

File: src/main.rs (hunk 10 above)
Comment (MahadMuhammad) [reply src/main.rs]: <details><summary>Yes, it produces error</summary>
<p>

image

</p>
</details>

Comment (MahadMuhammad) [reply src/main.rs]: Are you referring to this? Yes, it does compile :P
```suggestion
    try_parse()
```

Comment (CohenArthur) [reply src/main.rs]: yes, I was referring to the second snippet you posted :) I think it's cleaner to let anyhow handle all of that for us. can you commit that and then we'll merge it?

Comment (MahadMuhammad) [reply src/main.rs]: It's Updated now:)

I have been following this approach: https://github.com/chmln/sd/blob/master/src/main.rs#L22-L29. It only uses `main` for returning `anyhow` error code and `try_main` will handle all other things.

**Pull request merged**